### PR TITLE
Introduced sigma-data module

### PR DESCRIPTION
**Pull request by aslesarenko**
In this PR:

- [x] [AvlTreeVerifier into separate file](https://github.com/ScorexFoundation/sigmastate-interpreter/pull/924/commits/3e12acdc16f3df9d17db5e3ad93c9929691d248b)
- [x] [cleanup code + v6.0 todos](https://github.com/ScorexFoundation/sigmastate-interpreter/pull/924/commits/c3f4ef4535ada861611bc0b12c6fe904d16683c2)
- [x] [Value class moved to sigma.ast package](https://github.com/ScorexFoundation/sigmastate-interpreter/pull/924/commits/2d767ae75ab233deefeba25e42ca22ae22be8952)
- [x] [partially removed code related to TaggedVar](https://github.com/ScorexFoundation/sigmastate-interpreter/pull/924/commits/0ded24b7e3693184a659ce9d347c3982549f8492)
- [x] types, JitCost, CostKind, Operations, SMethod, trees, transformers, serializers moved to `sigma.ast` package of `data` module
- [x] [SparseArrayContainer.scala moved to sigma.utils](https://github.com/ScorexFoundation/sigmastate-interpreter/pull/924/commits/f3517edfa9700717c72ba966743380886d02a424)
- [x]  introduced ErgoTreeEvaluator, Profiler, AvlTreeVerifier interfaces
- [x] js.ErgoTree moved to sigma-data module
- [x] Added missing ScalaDocs

### CODE REVIEW

File: data/shared/src/main/scala/sigma/utils/Overloading.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package sigma.utils
+
+object Overloading {
```
Comment (kushti) [inline data/shared/src/main/scala/sigma/utils/Overloading.scala:3]: ScalaDocs missed

File: data/js/src/main/scala/sigma/ast/js/package.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package sigma.ast
+
+import sigma.Evaluation
+import sigma.ast.SType
```
Comment (kushti) [inline data/js/src/main/scala/sigma/ast/js/package.scala:4]: unused import

File: data/js/src/main/scala/sigma/ast/js/package.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import sigma.data.Iso
+import sigma.js.{Type, Value}
+import sigma.ast.Constant
```
Comment (kushti) [inline data/js/src/main/scala/sigma/ast/js/package.scala:7]: unused import

**Review by kushti — CHANGES_REQUESTED** [review]

File: data/shared/src/main/scala/sigma/ast/defs.scala [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+import org.ergoplatform.{ErgoBox, ErgoBoxCandidate}
+import sigma.kiama.rewriting.Rewriter.{everywherebu, rewrite, rule}
+import StdSigmaBuilder.mkUpcast
+import sigma.SigmaDataReflection
+import sigma.exceptions.InterpreterException
+
+import scala.annotation.nowarn
+import scala.reflect.classTag
+
+object defs {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply data/shared/src/main/scala/sigma/ast/defs.scala:15]: 1) Please suggest a name. 
2) Which exactly conventions not followed?

[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/ast/defs.scala:15]: Without ScalaDocs I can't really answer (1), on (2), naming could be ok actually if object is mimicking a package, but hard to say if this is the case, again, without description

**Comment by kushti** [comment]
Unused code after this PR: 

```
  /** Helper extractor to match SigmaProp values and extract ProveDlog out of it. */
  object ProveDlogProp {
    def unapply(p: SigmaProp): Option[ProveDlog] = SigmaDsl.toSigmaBoolean(p) match {
      case d: ProveDlog => Some(d)
      case _ => None
    }
  }
```

File: data/shared/src/main/scala/sigma/ast/defs.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply data/shared/src/main/scala/sigma/ast/defs.scala:15]: ok, renamed to `syntax` and added ScalaDocs

File: data/shared/src/main/scala/sigma/utils/Overloading.scala (hunk 1 above)
Comment (aslesarenko) [reply data/shared/src/main/scala/sigma/utils/Overloading.scala:3]: This class turns out to be duplicated, removed.

File: data/js/src/main/scala/sigma/ast/js/package.scala (hunk 2 above)
Comment (aslesarenko) [reply data/js/src/main/scala/sigma/ast/js/package.scala:4]: removed

File: data/js/src/main/scala/sigma/ast/js/package.scala (hunk 3 above)
Comment (aslesarenko) [reply data/js/src/main/scala/sigma/ast/js/package.scala:7]: removed

**Review by kushti — APPROVED** [review]

**Pull request merged**